src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/PackageSearchRunner.cs

```diff
@@ -0,0 +1,257 @@
+        private static Task<IEnumerable<IPackageSearchMetadata>> GetPackageAsync(
+            PackageSource source,
+            PackageSearchArgs packageSearchArgs,
+            CancellationToken cancellationToken)
+        {
+            return Task.Run(async () =>
+            {
+                using var cache = new SourceCacheContext();
+                var repository = Repository.Factory.GetCoreV3(source);
+                var resource = await repository.GetResourceAsync<PackageMetadataResource>(cancellationToken);
+
+                if (resource == null)
+                {
+                    return null;
+                }
+
+                return await resource.GetMetadataAsync(
+                    packageSearchArgs.SearchTerm,
+                    includePrerelease: packageSearchArgs.Prerelease,
+                    includeUnlisted: false,
+                    cache,
+                    packageSearchArgs.Logger,
+                    cancellationToken);
+            });
+        }
+
+        /// <summary>
+        /// Retrieves a list of package sources based on provided sources or default configuration sources.
+        /// </summary>
+        /// <param name="sources">The list of package sources provided.</param>
+        /// <param name="sourceProvider">The provider for package sources.</param>
+        /// <returns>A list of package sources.</returns>
+        private static IList<PackageSource> GetPackageSources(List<string> sources, IPackageSourceProvider sourceProvider)
+        {
+            IEnumerable<PackageSource> configurationSources = sourceProvider.LoadPackageSources()
+                .Where(p => p.IsEnabled);
+            IEnumerable<PackageSource> packageSources;
+
+            if (sources.Count > 0)
+            {
+                packageSources = sources
+                    .Select(s => PackageSourceProviderExtensions.ResolveSource(configurationSources, s));
+            }
+            else
+            {
+                packageSources = configurationSources;
+            }
+
+            return packageSources.ToList();
+        }
+
+        /// <summary>
+        /// Warns the user if the provided package sources use insecure HTTP connections.
+        /// </summary>
+        /// <param name="packageSources">The list of package sources to check.</param>
+        /// <param name="logger">The logger instance to use for logging.</param>
+        private static void WarnForHTTPSources(IList<PackageSource> packageSources, ILogger logger)
+        {
+            List<PackageSource> httpPackageSources = null;
+
+            foreach (PackageSource packageSource in packageSources)
+            {
+                if (packageSource.IsHttp && !packageSource.IsHttps && !packageSource.AllowInsecureConnections)
+                {
+                    if (httpPackageSources == null)
+                    {
+                        httpPackageSources = new();
```

Comment:
I think it's usually best to guess the capacity of the allocation.

```suggestion
                        httpPackageSources = new(capacity: packageSources.Count);
```